### PR TITLE
Add `--format` to `docker service ls`

### PR DESCRIPTION
**Pull request by yongtang**
**- What I did**

This fix tries to improve the display of `docker service ls` and adds `--format` flag to `docker service ls`.

In addition to `--format` flag, several other improvement:
1. Adds `--no-trunc` to `docker service ls`
2. Updates `docker stacks service`.
3. Adds `servicesFormat` to config file.

Related docs has been updated.

cc @thaJeztah @vdemeester @dnephin 

**- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
SGTM, can you add tests?

**Comment by vdemeester** [comment]
Design SGTM 👼

File: cli/command/service/list.go
Comment (thaJeztah) [inline cli/command/service/list.go]: I wonder if we should consider introducing `ShortID` and/or `TruncID` placeholders. The `TruncID` placeholder has the behavior of what `ID` has now (so, truncated by default, but show full ID if `--trunc` is added. That way people can define a template where the output _always_ has the full-id, or has an ID that can be truncated or full.

Comment (yongtang) [inline cli/command/service/list.go]: @thaJeztah does that mean we will leave `ID` alone regardless of `--no-trunc=false/true`, and only trunc `TruncID` based on `--no-trunc=true`?

Comment (thaJeztah) [inline cli/command/service/list.go]: @yongtang perhaps, yes. My initial thought was "do we need a `--no-trunc` flag? That's when I started thinking if we perhaps need other options for the format.

It's just a thought though, so perhaps @vdemeester or @dnephin has some thoughts

**Comment by yongtang** [comment]
@AkihiroSuda Added a test for the PR.

**Comment by thaJeztah** [comment]
Design LGTM, let's keep https://github.com/docker/docker/pull/28199#discussion_r87175650 for another discussion

**Comment by vdemeester** [comment]
ping @yongtang, needs a rebase 👼

**Comment by yongtang** [comment]
Thanks @vdemeester. The PR has been rebased.

**Comment by AkihiroSuda** [comment]
```
00:00:59.032 Errors from golint:
00:00:59.032 cli/command/service/list.go:94:1: comment on exported function GetServicesStatus should be of the form "GetServicesStatus ..."
00:00:59.032 
00:00:59.032 Please fix the above errors. You can test via "golint" and commit the result.
```

File: cli/command/service/list.go
Comment (vdemeester) [inline cli/command/service/list.go]: Same though as @thaJeztah, I don't really like `--no-trunc` flag, especially since format uses go template. We could provide a function that shorten "or not" the ID : `--format="{{ .ID | shorten }}"` or something like that.

File: cli/command/service/list.go
Comment (vdemeester) [inline cli/command/service/list.go]: This `else` is empty right ? should go 👼

File: cli/command/service/list.go
Comment (vdemeester) [inline cli/command/service/list.go]: Should/Could this return a map of struct instead, with mode and replica in the struct, instead of two maps ? 👼

**Review by vdemeester — CHANGES_REQUESTED** [review]
Few comments 👼 
I do agree with @thaJeztah on the `--no-trunc` flags (on other commands as well but 😅)

**Comment by yongtang** [comment]
@vdemeester @thaJeztah Thanks for the review. The PR has been updated and now `--no-trunc` are replace with template func of `shorten`. Please take a look and let me know if there are any issues.

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

File: integration-cli/docker_cli_swarm_test.go
Comment (AkihiroSuda) [inline integration-cli/docker_cli_swarm_test.go]: I wonder this test can be moved to UT, but it can be another PR in the future

File: docs/reference/commandline/service_ls.md
Comment (AkihiroSuda) [inline docs/reference/commandline/service_ls.md]: I wonder we should have two separate placeholders e.g. `RunningReplicas` and `AllReplicas`.

WDYT?

File: integration-cli/docker_cli_swarm_test.go
Comment (vdemeester) [inline integration-cli/docker_cli_swarm_test.go]: Yes 👍

File: docs/reference/commandline/service_ls.md
Comment (yongtang) [inline docs/reference/commandline/service_ls.md]: @AkihiroSuda I feel like that means we introduce new concepts that does not necessary match the existing Table header. Should we change the table header to `RUNNING/ALL REPLICAS`? Or maybe we could leave it alone for now?

**Comment by yongtang** [comment]
@AkihiroSuda @vdemeester Thanks for the review. The PR has been updated with integration tests have been replaced by unit tests. Please take a look.

File: docs/reference/commandline/service_ls.md
Comment (AkihiroSuda) [inline docs/reference/commandline/service_ls.md]: I think it is ok to introduce all `.Replicas`, `.RunningReplicas`, and `.AllReplicas`.
But it is ok to leave it alone for now.

**Review by AkihiroSuda — APPROVED** [review]
LGTM

PTAL @vdemeester @thaJeztah for confirmation.

**Comment by thaJeztah** [comment]
Let me put it in code-review for a bit, I'd like to confer with other maintainers on the `shorten` approach from a UX consistency perspective. I think it's useful, but need to be sure before we merge :heart:

**Comment by yongtang** [comment]
@thaJeztah in case we want to stick with `shorten` we may want to consider deprecated `--no-trunc` for other commands (for consistency). Either is fine for me though.

**Comment by thaJeztah** [comment]
@yongtang hm, okay discussed this in the maintainers meeting, and we think it would be better to put the `shorten` option in a separate PR, and make this PR just add the regular "formatting" option, so we can discuss the `shorten` option a bit more in depth

(sorry :smile:)

**Comment by yongtang** [comment]
Thanks @thaJeztah. The PR has been updated. Also created another PR #30484 for `shorten`. Please take a look.

Also in this PR I didn't add the `--no-trunc` flag and the output of ID will be truncated by default (same as old behavior). Please let me know if we want to introduce `--no-trunc` in this PR, or we want to have the full ID output by default (behavior will be different from original).

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Comment by vdemeester** [comment]
Moving to docs-review
/cc @thaJeztah

**Review by vdemeester — CHANGES_REQUESTED** [review]
@yongtang  Could you update `docs/reference/commandline/cli.md` to add an example for `ServicesFormat` ?

**Comment by yongtang** [comment]
Thanks @vdemeester. The docs has been updated in the PR. Please take a look and let me know if there are any issues.

**Review by vdemeester — APPROVED** [review]
docs LGTM 🐸
/cc @thaJeztah for docs revisit

**Pull request merged**
